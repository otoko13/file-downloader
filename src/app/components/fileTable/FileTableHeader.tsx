import classNames from "classnames";
import { FunctionComponent } from "react";
import Checkbox, { CheckedState } from "../checkbox/Checkbox";
import styles from "./fileTableHeader.module.scss";

export interface FileTableHeaderProps {
  checkedState: CheckedState;
  onSelectAllClicked: () => void;
  selectedFileCount: number;
  onDownloadClicked: () => void;
}

const FileTableHeader: FunctionComponent<FileTableHeaderProps> = ({
  checkedState,
  onSelectAllClicked,
  selectedFileCount,
  onDownloadClicked,
}) => {
  return (
    <thead className={styles["file-table-header"]}>
      <tr>
        <th colSpan={5}>
          <div className={styles["top-header"]}>
            <Checkbox
              ariaLabel="Select all"
              checkedState={checkedState}
              onClick={onSelectAllClicked}
            />
            <div className={styles["selected-count"]}>
              Selected {selectedFileCount}
            </div>
            <button
              className={styles["download-button"]}
              onClick={onDownloadClicked}
            >
              <i
                className={classNames("bi-download", styles["download-icon"])}
              />
              Download Selected
            </button>
          </div>
        </th>
      </tr>
      <tr>
        <th />
        <th>Name</th>
        <th>Device</th>
        <th>Path</th>
        <th className={styles["status-header"]}>Status</th>
      </tr>
    </thead>
  );
};
export default FileTableHeader;
