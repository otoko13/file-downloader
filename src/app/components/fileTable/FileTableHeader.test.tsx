import { render, screen } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import "@testing-library/jest-dom";
import FileTableHeader, { FileTableHeaderProps } from "./FileTableHeader";

let defaultProps: FileTableHeaderProps;
let downloadClickHandler: jest.Mock;
let selectAllClickHandler: jest.Mock;

beforeEach(() => {
  downloadClickHandler = jest.fn();
  selectAllClickHandler = jest.fn();
  defaultProps = {
    checkedState: "unchecked",
    onDownloadClicked: downloadClickHandler,
    onSelectAllClicked: selectAllClickHandler,
    selectedFileCount: 0,
  };
});

describe("appearance", () => {
  it("should match the snapshot", () => {
    const { container } = render(
      <table>
        <FileTableHeader {...defaultProps} />
      </table>
    );
    expect(container).toMatchSnapshot();
  });
});

describe("interactions", () => {
  it("should call onDownloadClicked when download button is clicked", async () => {
    render(
      <table>
        <FileTableHeader {...defaultProps} />
      </table>
    );

    expect(downloadClickHandler).not.toHaveBeenCalled();
    await userEvent.click(screen.getByRole("button"));
    expect(downloadClickHandler).toHaveBeenCalled();
  });

  it("should call onSelecteAllClicked when checkbox is clicked", async () => {
    render(
      <table>
        <FileTableHeader {...defaultProps} />
      </table>
    );

    expect(selectAllClickHandler).not.toHaveBeenCalled();
    await userEvent.click(screen.getByRole("checkbox"));
    expect(selectAllClickHandler).toHaveBeenCalled();
  });
});

describe("interactions", () => {
  it("should show the selected file count correctly", () => {
    render(
      <table>
        <FileTableHeader {...defaultProps} selectedFileCount={3} />
      </table>
    );
    expect(screen.getByText("Selected 3")).toBeInTheDocument();
  });
});
