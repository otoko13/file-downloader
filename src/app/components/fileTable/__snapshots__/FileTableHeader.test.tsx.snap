// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`appearance should match the snapshot 1`] = `
<div>
  <table>
    <thead
      class="file-table-header"
    >
      <tr>
        <th
          colspan="5"
        >
          <div
            class="top-header"
          >
            <button
              aria-checked="false"
              aria-label="Select all"
              class="checkbox"
              role="checkbox"
            />
            <div
              class="selected-count"
            >
              Selected 
              0
            </div>
            <button
              class="download-button"
            >
              <i
                class="bi-download download-icon"
              />
              Download Selected
            </button>
          </div>
        </th>
      </tr>
      <tr>
        <th />
        <th>
          Name
        </th>
        <th>
          Device
        </th>
        <th>
          Path
        </th>
        <th
          class="status-header"
        >
          Status
        </th>
      </tr>
    </thead>
  </table>
</div>
`;
